gobblin-compaction/src/main/java/gobblin/compaction/source/CompactionSource.java

```diff
@@ -18,32 +18,39 @@
 import gobblin.runtime.JobState;
 import gobblin.runtime.task.TaskUtils;
 import gobblin.source.Source;
+import gobblin.source.WorkUnitStreamSource;
 import gobblin.source.extractor.Extractor;
+import gobblin.source.workunit.BasicWorkUnitStream;
 import gobblin.source.workunit.WorkUnit;
+import gobblin.source.workunit.WorkUnitStream;
 import gobblin.util.HadoopUtils;
 import lombok.extern.slf4j.Slf4j;
 import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.fs.FileStatus;
 import org.apache.hadoop.fs.FileSystem;
 import org.apache.hadoop.fs.LocalFileSystem;
 import org.apache.hadoop.fs.Path;
-
 import java.io.IOException;
 import java.net.URI;
-import java.util.List;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
 import java.util.UUID;
+import java.util.concurrent.*;
+import java.util.List;
 
 /**
  * A compaction source derived from {@link Source} which uses {@link DefaultFileSystemGlobFinder} to find all
  * {@link Dataset}s. Use {@link CompactionSuite#getDatasetsFinderVerifiers()} to guarantee a given dataset has passed
  * all verification. Each found dataset will be serialized to {@link WorkUnit} by {@link CompactionSuite#save(Dataset, State)}
  */
 @Slf4j
-public class CompactionSource implements Source<String, String> {
+public class CompactionSource implements WorkUnitStreamSource<String, String> {
   private CompactionSuite suite;
   private Path tmpJobDir;
   private FileSystem fs;
+  private ExecutorService service;
 
+  @Override
   public List<WorkUnit> getWorkunits(SourceState state) {
```

Comment:
You should remove `getWorkUnits(SourceState)`, it is better to only have to maintain one functionally equivalent method.